src/private.rs

```diff
@@ -0,0 +1,256 @@
+            }
+        }
+
+        pub fn notify(&self) -> bool {
+            self.notified.store(true, Ordering::SeqCst);
+
+            if let Some(waker) = self.waker.take() {
+                waker.wake();
+
+                true
+            } else {
+                false
+            }
+        }
+
+        pub fn clear(&self) {
+            self.notified.store(false, Ordering::SeqCst);
+        }
+
+        #[allow(unused)]
+        pub fn wait(&self) -> impl Future<Output = ()> + '_ {
+            core::future::poll_fn(move |cx| self.poll_wait(cx))
+        }
+
+        pub fn poll_wait(&self, cx: &Context<'_>) -> Poll<()> {
+            self.waker.register(cx.waker());
+
+            if self.notified.swap(false, Ordering::SeqCst) {
+                Poll::Ready(())
+            } else {
+                Poll::Pending
+            }
+        }
+    }
+
+    // pub struct MultiNotificationInner<const N: usize> {
+    //     waker: MultiWakerRegistration<N>,
+    //     notified: bool,
+    // }
+
+    // pub struct MultiNotification<const N: usize> {
+    //     cs: IsrCriticalSection,
+    //     inner: UnsafeCell<MultiNotificationInner<N>>,
+    // }
+
+    // impl<const N: usize> MultiNotification<N> {
+    //     pub const fn new() -> Self {
+    //         Self {
+    //             cs: IsrCriticalSection::new(),
+    //             inner: UnsafeCell::new(MultiNotificationInner {
+    //                 waker: MultiWakerRegistration::new(),
+    //                 notified: false,
+    //             }),
+    //         }
+    //     }
+
+    //     pub fn notify(&self) -> bool {
+    //         let _cs = self.cs.enter();
+
+    //         let inner = unsafe { self.inner.get().as_mut() }.unwrap();
+
+    //         inner.notified = true;
+
+    //         inner.waker.wake()
+    //     }
+
+    //     pub fn wait(&self) -> impl Future<Output = ()> + '_ {
+    //         core::future::poll_fn(move |cx| self.poll_wait(cx))
+    //     }
+
+    //     pub fn poll_wait(&self, cx: &Context<'_>) -> Poll<()> {
+    //         let _cs = self.cs.enter();
+
+    //         let inner = unsafe { self.inner.get().as_mut() }.unwrap();
+
+    //         inner.waker.register(cx.waker()).unwrap();
+
+    //         if inner.notified {
+    //             Poll::Ready(())
+    //         } else {
+    //             Poll::Pending
+    //         }
+    //     }
+    // }
+
+    // unsafe impl<const N: usize> Send for MultiNotification<N> {}
+    // unsafe impl<const N: usize> Sync for MultiNotification<N> {}
+}
+
+pub mod completion {
+    use core::{
+        future::Future,
+        pin::Pin,
+        task::{Context, Poll},
+    };
+
+    pub struct Completion<F, D>
+    where
+        D: FnMut(bool),
+    {
+        fut: F,
+        destr: D,
+        completed: bool,
+    }
+
+    impl<F, D> Completion<F, D>
+    where
+        D: FnMut(bool),
+    {
+        pub const fn new(fut: F, destr: D) -> Self {
```

Comment:
It's probably safest if this accepts a pinned future instead of taking ownership, then you can remove the `unsafe` usage in the poll implementation.
```suggestion
        pub const fn new(fut: Pin<&mut F>, destr: D) -> Self {
```

Reply:
Also `dtor` is the common short name for destructor.

Reply:
Can you show a code snippet that shows how to get rid of the unsafe? I didn't know that's actually possible (but please no `pin_project` and such 3rd parties).

Reply:
```Rust
let future = pin!(async { notification.wait() });
Completion::new(future, ...).await;
```

```Rust
    impl<F, D> Future for Completion<F, D>
    where
        F: Future,
        D: FnMut(bool),
    {
        type Output = F::Output;

        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
            let fut = self.fut.clone();

            let poll = fut.poll(cx);

            this.completed |= matches!(&poll, Poll::Ready(_));

            poll
        }
    }
```

Reply:
Does not compile. I really do not understand how you expect to access `self.fut` which is inside the `Pin` without any unsafe.

Ditto for `this.completed`.

Reply:
And `clone` is even stranger in this context ?!?!

Reply:
I use clone because `poll` consumes the pin.

```Rust
    impl<F, D> Future for Completion<F, D>
    where
        F: Future,
        D: FnMut(bool),
    {
        type Output = F::Output;

        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
            let fut = self.as_ref().fut.clone();

            let poll = fut.poll(cx);

             // Change completed to a Cell<bool>
            let completed = &self.as_ref().completed;
            completed.set(completed.get() | matches!(&poll, Poll::Ready(_));

            // This should be possible as well since Self should be Unpin.
            self.as_mut().completed |= matches!(&poll, Poll::Ready(_));

            poll
        }
    }
```

Reply:
None of this compiles. I would gladly accept a solution that removes the `unsafe`, but you have to write it. Really.

Reply:
Fine. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=43894088765c3e036d452fde3be44171

Reply:
This has nothing to do with your original suggestion claiming that `Completion::poll` can be implemented in a safe way.

So I would appreciate if you are a tad more explicit and if something turns out to be impossible - to admit it, rather than me continuing to feel perplexed as to what is going on and am I misunderstanding how `Pin` is supposed to work or not.

Otherwise, the new solution is really neat and I'll take it. In fact is is absolutely in sync with what I'm doing so far (avoid writing explicit `Future` implementations by all means, and I'm a bit envious I did not come up with this as well!)